### PR TITLE
Refactor the workspace into the LoopBack 2.x structure

### PR DESCRIPTION
**Pull request by ritch**
This patch upgrades the workspace application into the LoopBack 2.x structure. The basic process was:

 - write a migration script that read the models.json and JS file and outputed 2.x style model files
 - added a "ready" event to allow the application to be booted enitrely before referencing models
 - removed some hacks for bugs fixed in boot 2.x (or by the ready event)
 - added a beforeEach reset for REST api tests

/to @raymondfeng

Connect to strongloop/loopback-workspace#213

### CODE REVIEW

**Comment by raymondfeng** [comment]
The code LGTM. We seem to miss a dependency on `compression`. 
```
Wed, 15 Jul 2015 20:55:34 GMT loopback:boot:config-loader merged test middleware configuration {"initial:before":{"loopback#favicon":{}},"initial":{"compression":{},"cors":{"params":{"origin":true,"credentials":true,"maxAge":86400}}},"session":{},"auth":{},"parse":{},"routes":{},"files":{},"final":{"loopback#urlNotFound":{}},"final:after":{"errorhandler":{}}}
Wed, 15 Jul 2015 20:55:34 GMT loopback:boot:compiler Skipping undefined - module not found
/mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:355
    throw err;
          ^
Error: Cannot resolve path "compression"
    at resolveAppPath (/mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:353:15)
    at resolveAppScriptPath (/mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:678:22)
    at resolveMiddlewarePath (/mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:554:20)
    at /mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:472:24
    at Array.forEach (native)
    at /mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:471:18
    at Array.forEach (native)
    at /mnt/jenkins/workspace/loopback-workspace.pr/4263f9af/node_modules/loopback-boot/lib/compiler.js:466:30
    at Array.forEach (native)
```

**Comment by ritch** [comment]
> We seem to miss a dependency on compression.

This looks like a dependency of an app generated to test the e2e functionality of the workspace...  I don't get this error locally. I think `strong-cache-install` might be related somehow.

**Comment by ritch** [comment]
test please

**Comment by rmg** [comment]
I get the same failure locally as CI does, every time.

    npm cache clear && git clean -fxd && npm install && npm test

**Comment by rmg** [comment]
The failure is not in the e2e tests, it is in any test that has `require('../server/server');`

The cause is you added a `server/middleware.json` file where there previously wasn't one. Now then the workspace itself is booted, it tries to load middleware for loopback-workspace - not a sample app - and it fails to find it.

File: server/middleware.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+  "initial:before": {
+    "loopback#favicon": {}
+  },
+  "initial": {
+    "compression": {},
```
Comment (rmg) [inline server/middleware.json]: Is there a reason loopback-workspace itself needs any middleware? If so, they'll need to be added as dependencies.

**Comment by ritch** [comment]
test please

**Comment by rmg** [comment]
@ritch is this change backwards compatible? If not, you should change the major version as part of the PR, otherwise once this merges everything that depends on loopback-workspace will fail on CI.

**Comment by ritch** [comment]
> @ritch is this change backwards compatible? If not, you should change the major version as part of the PR, otherwise once this merges everything that depends on loopback-workspace will fail on CI.

It should be. The changes are in the way the models are booted. The actual APIs should remain the same. The tests did not need to be change, so I am somewhat confident that consumers (eg. arc) will be fine. This is a fairly large change though, so I can't be 100% certain.

If you see anything that looks like a breaking change LMK. We shouldn't need to break anything to restructure the models / upgrade to `loopback-boot` 2.0.

**Comment by rmg** [comment]
> This is a fairly large change though, so I can't be 100% certain.

I didn't look very hard at the actual changes, it was the size that made me wonder.

**Comment by raymondfeng** [comment]
LGTM

**Pull request merged**